### PR TITLE
global: major refactoring

### PR DESCRIPTION
**Pull request by lnielsen**

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -24,13 +24,8 @@
 Changes
 =======
 
-Version 0.3.1 (released 2015-10-06)
```
Comment (jirikuncar) [inline CHANGES.rst]: it's not a big issue but probably we should keep the changes inplace.

Comment (lnielsen) [inline CHANGES.rst]: Agree. Just didn't clean this up yet.

File: invenio_base/app.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    Note that Invenio-Config provides a default configuration loader which is
+    sufficient for most cases.
+
+    Example of a WSGI factory:
+
+    .. code-block:: python
+
+       def my_wsgi_factory(app):
+           return DispatcherMiddleware(app.wsgi_app, {'/api': api_app})
+
```
Comment (nharraud) [inline invenio_base/app.py]: Missing `:return:` in doc

File: invenio_base/app.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        return app
+
+    return _create_app
+
+
+def create_cli(create_app=None):
+    """Create CLI for ``inveniomanage`` command.
+
+    :param create_app: Application factory.
```
Comment (nharraud) [inline invenio_base/app.py]: Missing :return: in doc

File: invenio_base/app.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+                init_func(ep.load())
+    if modules:
+        for m in modules:
+            init_func(m)
+
+
+def base_app(app_name, env_prefix='APP', instance_path=None,
+             static_folder=None, static_url_path='/static',
+             template_folder='templates'):
+    """Invenio base application factory."""
```
Comment (nharraud) [inline invenio_base/app.py]: Missing `:return:` in doc

File: invenio_base/__init__.py [hunk 5]
```diff
@@ -3,18 +3,77 @@
...
+   $ cd mysite
+   $ python manage.py run
+"""
+
+from __future__ import absolute_import, print_function
+
+from .version import __version__
 
-"""Build the Invenio application."""
+__all__ = ('__version__', )
```
Comment (nharraud) [inline invenio_base/__init__.py]: Maybe add `create_app_factory` to `__all__`?

**Comment by nharraud** [comment]
Reviewed, mainly little details.
Note for the future: as Invenio-base and the cookiecutter template will evolve and users won't be able to update their generated projects, it would be nice to have an example showing how to use create_app_factory with a basic use-case (one module, api and ui) in invenio-base. That way people can just read the diff to see how the example evolved. Diffing cookiecutter code or the code it generates is not that easy.

**Comment by nharraud** [comment]
Everything seems ok. :shipit: ?

**Pull request merged**